Framework/DataObjects/src/EventList.cpp

```diff
@@ -2162,6 +2263,117 @@ void EventList::generateCountsHistogram(const MantidVec &X, MantidVec &Y) const
   } // end if (there are any events to histogram)
 }
 
+/** Find the bin which this TOF value falls in with linear binning, assumes TOF is in range of X
+ *
+ * bin_number =  (tof - xmin) / step
+ *
+ * @param X :: The x bins
+ * @param tof :: TOF of the event we are trying to bin
+ * @param divisor :: pre-calculated divisor
+ * @param offset :: pre-calculated offset
+ */
+boost::optional<size_t> EventList::findLinearBin(const MantidVec &X, const double tof, const double divisor,
+                                                 const double offset) {
+  return findExactBin(X, tof, static_cast<size_t>(tof / divisor - offset));
+}
+
+/** Find the bin which this TOF value falls in with log binning, assumes TOF is in range of X
+ *
+ * formula to get bin_number from TOF
+ *
+ * bin_number = log_{abs(step) + 1}(tof/xmin) (that is log base abs(step)+1)
+ *
+ * rewritten as
+ *
+ * bin_number = log(tof)/log(abs(step)+1) - log(xmin)/log(abs(step)+1)
+ *
+ * or bin_number = log(tof)/divisor - offset, where divisor and offset are precalulated
+ *
+ * @param X :: The x bins
+ * @param tof :: TOF of the event we are trying to bin
+ * @param divisor :: pre-calculated divisor
+ * @param offset :: pre-calculated offset
+ */
+boost::optional<size_t> EventList::findLogBin(const MantidVec &X, const double tof, const double divisor,
+                                              const double offset) {
+  return findExactBin(X, tof, static_cast<size_t>(log(tof) / divisor - offset));
+}
+
+/** Find the exact bin which a TOF falls in starting from the provided estimated one.
+ *
+ * @param X :: The x bins
+ * @param tof :: TOF of the event we are trying to bin
+ * @param n_bin :: starting estiamted bin number
+ */
+boost::optional<size_t> EventList::findExactBin(const MantidVec &X, const double tof, size_t n_bin) {
```

Comment:
I feel like adding a check just before the while loops that the tof is already in the right bin would speed this up. Unfortunately, I don't know how often the calling code gets the right bin from the start
```c++
if (tof >= X[n_bin] && tof < X[n_bin+1])
  return n_bin;
```
would already be covered by the `while` loops, but doesn't require setting up a while loop. I don't have numbers, but for event splitting protecting the while loops by an `if` like this makes a noticeable difference.

Reply:
Since all the bin estimates should land with 1 bin I changed this to a `if`/`else if` which reduce the number of operations needed. Although I don't see any performance improvement.